File: 05-java-mqtt/src/main/java/demo/ProtelisConfigSpec.java
package demo;

import com.uchuhimo.konf.ConfigSpec;
import com.uchuhimo.konf.RequiredItem;
import demo.data.MqttProtelisNode;

import java.util.List;

/**
 * Konf specification for the TOML configuration file.
 */
@SuppressWarnings("PMD.AvoidDuplicateLiterals")
public final class ProtelisConfigSpec {

    /**
     * Create a new config specification named protelis.
     */
    public static final ConfigSpec SPEC = new ConfigSpec("protelis");
    /**
     * Required field iterations. It represents the number of cycles the simulation will perform.
     */

    @SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<Integer> iterations =
            new RequiredItem<Integer>(SPEC, "iterations") { };
    /**
     * Required field protelisModuleName. It is the name of the .pt source file.
     */
    @SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<String> protelisModuleName =
            new RequiredItem<String>(SPEC, "protelisModuleName") { };

    /**
     * Required field brokerHost. It is the IP address of the MQTT broker.
     */
    @SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<String> brokerHost =
            new RequiredItem<>(SPEC, "brokerHost") { };

    /**
     * Required field brokerPort. It is the port of the MQTT broker.
     */
    @SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<Integer> brokerPort =
            new RequiredItem<>(SPEC, "brokerPort") { };

    /**
     * Required field nodes. It contains a list of protelis nodes.
     */
    @SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<List<MqttProtelisNode>> nodes =
            new RequiredItem<List<MqttProtelisNode>>(SPEC, "nodes") { };

    private ProtelisConfigSpec() { }
}

File: 06-kotlin-mqtt/src/main/kotlin/org/protelis/demo/ProtelisConfigSpec.kt
package org.protelis.demo

import com.uchuhimo.konf.ConfigSpec

/**
 * A Protelis node with [id] listening to a certain [listen] address and with a pre-defined set of [neighbors].
 * May or may not be [leader] (defaults to `false`)
 */
data class ProtelisNode(
    val id: Int,
    val listen: String,
    val neighbors: Set<String>,
    val leader: Boolean = false,
)

/**
 * This is used by Konf to map the TOML main configuration to a class. Includes:
 * - number of [iterations] to perform;
 * - [protelisModuleName] to execute;
 * - [brokerHost] address and [brokerPort];
 * - list of [nodes] composing the system.
 */
object ProtelisConfigSpec : ConfigSpec("protelis") {
    /**
     * Number of iterations to perform.
     */
    val iterations by required<Int>(description = "Number of iterations the nodes have to run.")

    /**
     * Protelis module to execute.
     */
    val protelisModuleName by required<String>()

    /**
     * Host address.
     */
    val brokerHost by required<String>()

    /**
     * Host port.
     */
    val brokerPort by required<Int>()

    /**
     * Nodes in the system.
     */
    val nodes by required<List<ProtelisNode>>(description = "List of nodes")
}
